book/src/04_3_3_i2c.md

```diff
@@ -1 +1,187 @@
-# I2C
+# Writing an I²C Driver
+
+We're not going to write an entire driver, merely the first step: the `hello world` of driver writing: Reading the device ID of the sensor. This version is labelled hard, because you have to come up with the content of the methods and research information in the embedded-hal and data-sheets yourself. 
+
+`i2c-driver/src/icm42670p.rs` is a gap text of a very basic I²C IMU sensor driver. The task is to complete the file, so that running `main.rs` will log the device ID of the driver. The this gap text driver is based on the version of the same name that lives in common, but provides a little bit more functionality.
+
+`i2c-driver/src/icm42670p_solution.rs` provides the solution to this exercise. If you want to run it, the imports need to be changed in `main.rs` and `lib.rs`. The imports are already there, you only need to comment the current imports out and uncomment the solutions as marked in the line comments. 
+
+## Driver API
+
+### Instance of the Sensor
+
+ ✅ Create a struct that represents the sensor. It has two fields, one that represents the sensor's device address and one that represents the `I²C` bus itself. This is done using traits defined in the `embedded-hal` crate. The struct is public as it needs to be accessible from outside this crate, but its fields are private. 
+
+
+<Details>
+    <Summary>Example Solution</Summary>
+```rust 
+#[derive(Debug)]
+pub struct ICM42670P<I2C> {
+    /// The concrete I²C device implementation.
+    i2c: I2C,
+
+    /// Device address
+    address: DeviceAddr,
+}
+
+```
+</Details>
+
+
+ ✅ Implement an instantiating method in the `impl` block. This method needs to be accessible from outside, so it's labelled `pub`. The method takes ownership of the I²C bus and creates an instance of the struct you defined earlier.
+
+<Details>
+    <Summary>Example Solution</Summary>
+```rust
+impl<I2C, E>ICM42670P<I2C>
+where
+    I2C: i2c::WriteRead<Error = E> + i2c::Write<Error = E>,
+{
+    /// Create a new instance of the ICM42670P.
+    pub fn new(i2c: I2C, address: DeviceAddr) -> Result<Self, E> {
+
+        let icm42670p = ICM42670P { i2c, address };
```

Comment:
It looks like this in `lib.rs`. Here shouldn't it be only `Ok(Self { i2c, address })`?